### DISCUSSION TITLE
How to debug Next.js components?

### DISCUSSION THREAD

**Discussion by heloineto** [vercel/next.js#58486] [Help]
### Summary

Given a component, I want to know:

- Is it a Server Component? (at runtime)
- Is it a Client Component? (at runtime)
- What rendering strategy was used for the component? (SSG, SSR, ISR, CSR, PPR)

Is there any way to do this? A devtool perhaps?

**Comment by icyJoseph** [comment]
If it is a client component, then there's JavaScript for it. So you won't necessarily be able to debug Server Components, on the client.

There's however, a tool you can use:

- Tool: https://rsc-parser.vercel.app/
- Source code: https://github.com/alvarlagerlof/rsc-parser

The idea being that you are able to see, what Client Side slots the serialized VDOM contains, including references to the JS that drives the component, and also see which parts resolved to an HTML string.

<img width="1230" alt="Screenshot 2023-11-15 at 15 26 51" src="https://github.com/vercel/next.js/assets/21013447/dbcb95d6-7bab-4797-bf7c-0f7a7f042513">

This might also be worth reading: https://demystifying-rsc.vercel.app/static-content/2/
